### PR TITLE
Prepare dependents rules for 2021 computation while retaining current 2020 computation

### PR DESCRIPTION
**Pull request by alaroia-cfa**

### CODE REVIEW

File: app/models/dependent.rb [hunk 1]
```diff
@@ -252,6 +185,29 @@ def mixpanel_data
...
+  # Methods on Dependent::Rules can be accessed (and mocked-out) as yr_2020_* and yr_2021_*. In the future, we might
+  # add a default year with no prefix.
+  delegate :age, :born_in_last_6_months?, :disqualified_child_qualified_relative?, :meets_qc_age_condition?, :meets_qc_residence_condition?, :qualifying_child?, :qualifying_relative?, to: :rules_2020, prefix: :yr_2020
+  delegate :age, :born_in_last_6_months?, :disqualified_child_qualified_relative?, :meets_qc_age_condition?, :meets_qc_residence_condition?, :qualifying_child?, :qualifying_relative?, to: :rules_2021, prefix: :yr_2021
+
+  private
+
+  def rules_default
+    rules(intake.most_recent_filing_year)
+  end
```
Comment (jenny-heath) [inline app/models/dependent.rb:197]: reminder: this is unused

File: app/models/dependent/rules.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def qualifying_child?
+    @qualifying_child_relationship &&
+      @ssn_present &&
+      @meets_qc_claimant_condition &&
+      @meets_qc_misc_conditions &&
+      meets_qc_age_condition? &&
+      meets_qc_residence_condition? &&
+      age >= 0
```
Comment (jenny-heath) [inline app/models/dependent/rules.rb:51]: consider rolling this into the age condition method

File: spec/models/dependent/rules_spec.rb [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  let(:qualifying_child_relationship) { false }
+  let(:qualifying_relative_relationship) { false }
+  let(:ssn_present) { false }
+  let(:meets_misc_qualifying_relative_requirements) { false }
+  let(:meets_qc_residence_condition_generic) { false }
+  let(:meets_qc_claimant_condition) { false }
+  let(:meets_qc_misc_conditions) { false }
+  let(:subject) { described_class.new(birth_date, tax_year, full_time_student_yes, permanently_totally_disabled_yes, ssn_present, qualifying_child_relationship, qualifying_relative_relationship, meets_misc_qualifying_relative_requirements, meets_qc_residence_condition_generic, meets_qc_claimant_condition, meets_qc_misc_conditions) }
+
+  describe ".born_in_last_6_months?" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:16]: consider renaming to "final_6_months" to be clear that it's H2

File: spec/models/dependent/rules_spec.rb [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+    context "when born on June 30" do
+      let(:birth_date) { Date.new(tax_year, 6, 30) }
+
+      it "is true" do
+        expect(subject.born_in_last_6_months?).to be_truthy
+      end
+    end
+
+    context "when born on June 30" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:33]: This should be a test for actually late in the year, like December 30

File: spec/models/dependent/rules_spec.rb [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    context "when born on Jan 1 of the tax year" do
+      let(:birth_date) { Date.new(tax_year, 1, 1) }
+
+      it "is 0" do
+        expect(subject.age).to eq(0)
+      end
+    end
+
+    context "when born on Dec 31 of the previous year" do
+      let(:birth_date) { Date.new(tax_year - 1, 12, 1) }
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:52]: `12, 1` should be `12, 31` (or change the name)

File: spec/models/dependent/rules_spec.rb (hunk 3 above)
Comment (jenny-heath) [reply spec/models/dependent/rules_spec.rb]: Done!

File: spec/models/dependent/rules_spec.rb (hunk 4 above)
Comment (jenny-heath) [reply spec/models/dependent/rules_spec.rb]: Fixed!

File: spec/models/dependent/rules_spec.rb (hunk 5 above)
Comment (jenny-heath) [reply spec/models/dependent/rules_spec.rb]: Fixed!

File: spec/models/dependent/rules_spec.rb [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  describe ".disqualified_child_qualified_relative?" do
+    context "with a relationship that's normally a qualifying child" do
+      let(:qualifying_child_relationship) { true }
+      context "when young" do
+        let(:birth_date) { Date.new(tax_year - 2, 12, 25) }
+        it "is not a disqualified-child child relative" do
+          expect(subject.disqualified_child_qualified_relative?).to eq(false)
+        end
+      end
+      context "when old" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:113]: needs more newlines

File: spec/models/dependent/rules_spec.rb [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+          expect(subject.disqualified_child_qualified_relative?).to eq(false)
+        end
+      end
+    end
+  end
+
+  describe ".qualifying_relative?" do
+    context "with an old dependent who meets misc requirements and has a ssn/itin/atin stored" do
+      let(:meets_misc_qualifying_relative_requirements) { true }
+      let(:ssn_present) { true }
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:139]: we don't test this value but it's part of the method so we should add a case

File: spec/models/dependent/rules_spec.rb [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+            expect(subject.qualifying_relative?).to eq false
+          end
+        end
+      end
+
+      context "with a dependent who has a qualified relative relationship" do
+        let(:birth_date) { Date.new(tax_year - 30, 12, 25) }
+        let(:qualifying_relative_relationship) { true }
+
+        it "returns true" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:162]: `context "when misc qualifying relative requirements are met" do`

File: spec/models/dependent/rules_spec.rb [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        let(:birth_date) { Date.new(tax_year - 40, 12, 25) }
+        it "is not a disqualified-child child relative" do
+          expect(subject.disqualified_child_qualified_relative?).to eq(false)
+        end
+      end
+    end
+  end
+
+  describe ".qualifying_relative?" do
+    context "with an old dependent who meets misc requirements and has a ssn/itin/atin stored" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:137]: `    context "with an old dependent" do` `context "who meets misc requirements" do` `context "who has a ssn/itin/atin stored" do`

File: spec/models/dependent/rules_spec.rb [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+      let(:meets_qc_misc_conditions) { true }
+
+      context "when the child is < 6 months old" do
+        let(:birth_date) { Date.new(tax_year, 12, 25) }
+
+        it "returns true" do
+          expect(subject.qualifying_child?).to eq(true)
+        end
+
+        context "and not born yet" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:229]: i guess we could remove this now that the < 0 age condition is rolled into the age condition method

File: spec/models/dependent/rules_spec.rb [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+      context "when older than 6 months" do
+        let(:birth_date) { Date.new(tax_year, 1, 1) }
+        it "returns false" do
+          expect(subject.meets_qc_residence_condition?).to eq(false)
+        end
+      end
+    end
+  end
+
+  describe ".qualifying_child?" do
```
Comment (jenny-heath) [inline spec/models/dependent/rules_spec.rb:215]: i'm having trouble following this set of context blocks. they seem...random? the method basically just checks that everything in a list is true (the list consisting of some columns on dependent and some methods on rules) so i would have expected roughly:

```
context "when the child meets all of the qualifying conditions" do
end

context "when the child doesn't meet one or more of the qualifying conditions"do
end
```

...where each column / method was stubbed as true/false. the second block is trickier because it could feel excessively long to make one thing at a time false, but maybe something clever could be done looping over a list or something.
anyway, i don't get the age checks here, they seem to belong in `meets_qc_age_condition?`. i also realize this could be a holdover from the old code but maybe we should fix it

File: spec/models/dependent_spec.rb [hunk 12]
```diff
@@ -407,7 +234,7 @@
 
     it "returns the expected hash" do
       expect(dependent.mixpanel_data).to eq({
-        dependent_age_at_end_of_tax_year: "4",
+        dependent_age_at_end_of_tax_year: "5",
```
Comment (jenny-heath) [inline spec/models/dependent_spec.rb:237]: lol

File: app/models/intake.rb [hunk 13]
```diff
@@ -459,6 +455,10 @@ def opted_into_notifications?
     sms_notification_opt_in_yes? || email_notification_opt_in_yes?
   end
 
+  def tax_year
+    2019 # TODO: This is wrong. Migrate to a Rails setting or similar
+  end
```
Comment (jenny-heath) [inline app/models/intake.rb:460]: wait i thought you got rid of this...

File: bin/test [hunk 14]
```diff
@@ -6,5 +6,8 @@ if [ $(sysctl -n hw.logicalcpu) != "1" ] && ! psql "postgresql://localhost/vita-
   exit 1
 fi
 
+# Increase the # of max open files to avoid https://stackoverflow.com/questions/59890432/rack-error-runtimeerror-failed-to-get-urandom-in-a-rails-app-rails-5-0-6-ru
+ulimit -n 524288
```
Comment (jenny-heath) [inline bin/test:10]: if i were being really nitpicky i'd say to put this in a separate commit

Comment (alaroia-cfa) [reply bin/test]: Done! :) Removed that from this PR.

File: app/models/intake.rb (hunk 13 above)
Comment (alaroia-cfa) [reply app/models/intake.rb]: I somehow had missed it. Removed now.

**Comment by alaroia-cfa** [comment]
@jenny-heath I haven't made all the progress I wanted here, but let's sync up this morning. It'd be nice to work together on the test refactors. I also have a WIP on my laptop that makes some other changes.

File: app/controllers/ctc/questions/confirm_payment_controller.rb [hunk 15]
```diff
@@ -6,7 +6,7 @@ class ConfirmPaymentController < QuestionsController
       layout "intake"
 
       def edit
-        tax_return = current_intake.tax_return(2020)
+        tax_return = current_intake.default_tax_return
```
Comment (jenny-heath) [inline app/controllers/ctc/questions/confirm_payment_controller.rb:9]: move to ctc intake

File: app/models/intake.rb [hunk 16]
```diff
@@ -455,20 +451,8 @@ def opted_into_notifications?
...
-
-    tax_year - primary_birth_date.year
-  end
-
-  def spouse_age_end_of_tax_year
-    return unless spouse_birth_date.present?
-
-    tax_year - spouse_birth_date.year
+  def default_tax_return
+    tax_returns.find_by(year: Rails.application.config.default_tax_year)
```
Comment (jenny-heath) [inline app/models/intake.rb:455]: move to ctc intake

File: app/services/mixpanel_service.rb [hunk 17]
```diff
@@ -264,8 +264,8 @@ def data_from_intake(intake)
         intake_source: intake.source,
         intake_referrer: intake.referrer,
         intake_referrer_domain: intake.referrer_domain,
-        primary_filer_age_at_end_of_tax_year: intake.age_end_of_tax_year.to_s,
-        spouse_age_at_end_of_tax_year: intake.spouse_age_end_of_tax_year.to_s,
+        primary_filer_age_at_end_of_tax_year: age_from_date_of_birth(intake.primary_birth_date).to_s,
```
Comment (jenny-heath) [inline app/services/mixpanel_service.rb:267]: `age_at_end_of_tax_year`

File: config/application.rb [hunk 18]
```diff
@@ -50,5 +50,6 @@ class Application < Rails::Application
       production: "www.getctc.org"
     }
     config.middleware.use Middleware::CleanupMimeTypeHeaders
+    config.default_tax_year = 2020
```
Comment (jenny-heath) [inline config/application.rb:53]: consider renaming to `ctc_default_tax_year` since there isn't really a default year for gyr services

File: spec/factories/clients.rb [hunk 19]
```diff
@@ -53,7 +53,7 @@
         filing_status { "single" }
       end
       after(:create) do |client, evaluator|
-        create :tax_return, evaluator.status, year: 2020, client: client, filing_status: evaluator.filing_status
+        create :tax_return, evaluator.status, year: Rails.application.config.default_tax_year, client: client, filing_status: evaluator.filing_status
```
Comment (jenny-heath) [inline spec/factories/clients.rb:56]: we don't need to set the year to anything special in the factory

File: docs/2021-11-16-changing-the-year.md [hunk 20]
```diff
@@ -0,0 +1,36 @@
+# Changing the year
+
+Once a year or so, the team will need to adjust the app to handle the next year's taxes. Three aspects of this are
+covered below.
+
+## Configuring the default tax year
+
+To change the default `TaxReturn` year that we create during intake, change `Rails.application.config.default_tax_year`.
```
Comment (jenny-heath) [inline docs/2021-11-16-changing-the-year.md:8]: should become ctc specific
